lib/terraform/runner.rb

```diff
@@ -15,99 +15,99 @@ def available?
+      #   4. ActionType::RETRIEVE (Fetch)       fetch a existing stack object json from terraform-runner.
+      #   5. ActionType::CANCEL   (Stop)        stop running stack job in terraform-runner.
+      # * keys allowed in ':options'
+      #   - :input_vars                  [Hash]   (optional) key/value pairs, as input variables for the terraform-runner run job.
+      #   - :input_vars_type_constraints [Hash]   (optional) key/value(type constraints object, from Terraform Runner) pairs.
+      #   - :tags                        [Hash]   (optional) key/value pairs tags for terraform-runner Provisioned resources.
+      #   - :credentials                 [Array]  (optional) list of authentication objects for the terraform run job.
+      #   - :env_vars                    [Hash]   (optional) key/value pairs, used as environment variables, for terraform-runner run job.
+      #   - :name                        [String] (optional) name for new created stack in terraform-runner.
+      #   - :stack_id                    [String] [required] if Reconfigure/Retire/Retrieve/Cancel actions.
       #
-      # @return [Terraform::Runner::ResponseAsync] Response object of terraform-runner api call
-      def delete_stack(stack_id, template_path, input_vars: {}, input_vars_type_constraints: [], credentials: [], env_vars: {})
-        if stack_id.present? && template_path.present?
-          _log.debug("Run_aysnc/delete_stack('#{stack_id}') for template: #{template_path}")
-          input_params = ApiParams.to_normalized_cam_parameters(input_vars, input_vars_type_constraints)
-          response = delete_stack_job(
-            stack_id,
-            template_path,
-            :input_params => input_params,
-            :credentials  => credentials,
-            :env_vars     => env_vars
-          )
-          Terraform::Runner::ResponseAsync.new(response.stack_id)
-        else
-          _log.error("'stack_id' && 'template_path' are required for #{ResourceAction::RETIREMENT} action")
-          raise "'stack_id' && 'template_path' are required for #{ResourceAction::RETIREMENT} action"
-        end
+      def run(action_type, template_path, options = {})
+        raise "Not supported action type in this method, instead use method parse_terraform_variables" if action_type == ActionType::TEMPLATE_VARIABLES
+        raise "Not supported action type '#{action_type}'" unless ActionType.actions.include?(action_type)
+
+        _log.debug("Run #{action_type} for template: #{template_path}")
+
+        response = run_terraform_runner_stack_api(
+          Request.new(action_type)
+            .template_path(template_path)
+            .credentials(options[:credentials])
+            .input_vars(options[:input_vars], options[:input_vars_type_constraints])
+            .tenant_id(stack_tenant_id)
+            .tags(options[:tags])
+            .env_vars(options[:env_vars])
+            .name(options[:name])
+            .stack_id(options[:stack_id])
```

Comment:
Maybe it is just me but I had a hard time working out what was going on here :smile: 
Probably because these look like "getters" when they're really "setters", plus the return self and call the next method on the result of the previous.
Maybe in a follow-up this could be changed to something like
```
Request.new(action_type).tap do |req|
  req.template_path = template_path
  req.credentials = options[:credentials]
  ...
end
```
Or accept these as parameters in the `#initialize` method.